src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationFsm.java

```diff
@@ -0,0 +1,340 @@
+import org.openkilda.messaging.info.group.GroupDumpResponse;
+import org.openkilda.messaging.info.meter.MeterDumpResponse;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.SwitchId;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.rulemanager.RuleManagerConfig;
+import org.openkilda.rulemanager.RuleManagerImpl;
+import org.openkilda.wfm.CommandContext;
+import org.openkilda.wfm.error.FlowNotFoundException;
+import org.openkilda.wfm.error.IllegalFlowStateException;
+import org.openkilda.wfm.error.SwitchNotFoundException;
+import org.openkilda.wfm.share.metrics.MeterRegistryHolder;
+import org.openkilda.wfm.topology.flowhs.fsm.common.NbTrackableFlowProcessingFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.haflow.validation.HaFlowValidationFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.haflow.validation.HaFlowValidationFsm.State;
+import org.openkilda.wfm.topology.flowhs.service.FlowValidationEventListener;
+import org.openkilda.wfm.topology.flowhs.service.FlowValidationHubCarrier;
+
+import io.micrometer.core.instrument.LongTaskTimer;
+import io.micrometer.core.instrument.LongTaskTimer.Sample;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.squirrelframework.foundation.fsm.StateMachineBuilder;
+import org.squirrelframework.foundation.fsm.StateMachineBuilderFactory;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public final class HaFlowValidationFsm extends NbTrackableFlowProcessingFsm<HaFlowValidationFsm, State, Event, Object,
+        FlowValidationHubCarrier, FlowValidationEventListener> {
+    private static final String FINISHED_WITH_ERROR_METHOD_NAME = "finishedWithError";
+    private static final String FINISHED_METHOD_NAME = "finished";
+    @Getter
+    private final String flowId;
+    private final HaFlowValidationService service;
+    private int awaitingRulesSize;
+    private int awaitingMetersSize;
+    private int awaitingGroupsSize;
+    private final List<FlowDumpResponse> receivedRules = new ArrayList<>();
+    private final List<MeterDumpResponse> receivedMeters = new ArrayList<>();
+    private final List<GroupDumpResponse> receivedGroups = new ArrayList<>();
+    private final Set<SwitchId> switchIdSet = new HashSet<>();
+    private HaFlowValidationResponse response;
+
+    public HaFlowValidationFsm(@NonNull CommandContext commandContext, @NonNull FlowValidationHubCarrier carrier,
+                               @NonNull String flowId, @NonNull PersistenceManager persistenceManager,
+                               @NonNull Collection<FlowValidationEventListener> eventListeners,
+                               @NonNull RuleManagerConfig ruleManagerConfig) {
```

Comment:
Please do like you did in Flova validation fsm - pass here RuleManager ruleManager, not the config.
Let hub to decide which implementation we need here

Reply:
fixed